### ISSUE TITLE
Implement custom binary packing interface

### ISSUE THREAD

**Issue by theo-zil** [Zilliqa/zq2#539]
Make storing data to persistence easier:
 *  Allow appending to vectors without deserializing: specifically, for `touched_address_index`, we currently call `bincode` twice (deserialize to `Vec`, call `Vec::push()`, serialize back to `[u8]`). That adds a good bit of overhead.
 * Short-circuiting binary-representable types: integer types can be converted with `to_be_bytes()`, and our `Hash` has a directly accessible internal `[u8]` (also directly available through the `as_bytes()` method). Calling `bincode` seems to add extra overhead due to the indirection and genericness of the library.

In short, the idea is probably to make a trait that can be implemented for relevant types which will short-circuit for the types that allow it, implement the custom packing for vecs to allow appending, and probably just fallback to `bincode` for everything else that doesn't need a special case for now.
